### ISSUE TITLE
[salt-cloud] Minion doesn't install on some providers after upgrading to 2015.8.0

### ISSUE THREAD

**Issue by geraint-jones-nz** [saltstack/salt#26699]
Hi

I have recently updated our master to the following version:
```console
root@mgmt1:/srv/saltstack# salt --versions
Salt Version:
           Salt: 2015.8.0-316-g226ef73

Dependency Versions:
         Jinja2: 2.7.3
       M2Crypto: 0.21.1
           Mako: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.0.1
         Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.4
       dateutil: Not Installed
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
       pycrypto: 2.6.1
        timelib: Not Installed

System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 3.13.0-24-generic
         system: Ubuntu 14.04 trusty
```
Since doing this `salt-cloud` no longer installs the minion, the process completes without error but no minion is installed.

**Comment by jfindlay** [comment]
@geraint-jones-nz, thanks for the report.  Is it possible for you to provide the debug log output of the salt-cloud command you are using, obfuscating any private details if necessary?

**Comment by geraint-jones-nz** [comment]
It is very very large, here goes : https://github.com/geraint-jones-nz/salt/blob/develop/log.tgz

**Comment by jfindlay** [comment]
Thanks for the extra info, @geraint-jones-nz.  It seems there may be a problem with the bootstrapping or the interaction between that and salt-cloud.

**Comment by geraint-jones-nz** [comment]
This is becoming a blocker for us, is anyone willing to take a look ?

**Comment by jfindlay** [comment]
Ping @rallytime, @cro

**Comment by geraint-jones-nz** [comment]
Updated and its still present:

> root@MGT01:~# salt --versions
> Salt Version:
>            Salt: 2015.8.0-361-gc3a2355

**Comment by rallytime** [comment]
@geraint-jones-nz Can you go into further details about what you're seeing? Does this work if you try to use salt-cloud to provision only one instance on SoftLayer? (Just trying to narrow down if something is going wrong with using a map file or perhaps using `-P`.) Hopefully spinning up only one instance at a time should help narrow down a failure point.

Could you also post your cloud configs, being sure to remove any sensitive data?

I just tested spinning up a single minion on SoftLayer and I was able to get it working:
```
root@rallytime:~# salt-cloud -p sl-ubuntu-14 nt-sl-test-1
<snipped>
nt-sl-test-1:
    ----------
    accountId:
        <redacted>
    createDate:
        2015-09-10T22:38:20-05:00
    deployed:
        True
    domain:
        example.com
    fullyQualifiedDomainName:
        nt-sl-test-1.example.com
    globalIdentifier:
        <redacted>
    hostname:
        nt-sl-test-1
    id:
        <redacted>
    lastPowerStateId:
    lastVerifiedDate:
    maxCpu:
        1
    maxCpuUnits:
        CORE
    maxMemory:
        1024
    metricPollDate:
    modifyDate:
    password:
        <redacted>
    provisionDate:
    public_ip:
        50.23.124.234
    startCpus:
        1
    statusId:
        1001
    username:
        root
    uuid:
        <redacted>
```
And the minion responds to a ping successfully:
```
root@rallytime:~# salt nt-sl-test-1 test.ping
nt-sl-test-1:
    True
```
Versions Report, from a recent pull from today on the 2015.8 branch:
```
root@rallytime:~# salt-cloud --versions
Salt Version:
            Salt: 2015.8.0-4-g757d3c4

Dependency Versions:
 Apache Libcloud: 0.18.0
          Jinja2: 2.7.2
        M2Crypto: 0.21.1
            Mako: Not Installed
          PyYAML: 3.10
           PyZMQ: 14.5.0
          Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
            RAET: Not Installed
         Tornado: 4.2.1
             ZMQ: 4.0.5
            cffi: 0.8.6
        cherrypy: Not Installed
        dateutil: Not Installed
           gitdb: 0.6.4
       gitpython: Not Installed
           ioflo: Not Installed
         libnacl: Not Installed
    msgpack-pure: Not Installed
  msgpack-python: 0.3.0
    mysql-python: Not Installed
       pycparser: 2.10
        pycrypto: 2.6.1
          pygit2: Not Installed
    python-gnupg: Not Installed
           smmap: 0.9.0
         timelib: Not Installed

System Versions:
            dist: Ubuntu 14.04 trusty
         machine: x86_64
         release: 4.1.5-x86_64-linode61
          system: Ubuntu 14.04 trusty
```

**Comment by geraint-jones-nz** [comment]
Hi.

It does all appear to work, however salt is never installed.

This is the `cloud.profile` entry for this profile

```
web_proxy_mel_sl:
  provider: 9s-softlayer
  image: UBUNTU_14_64
  cpu_number: 2
  ram: 4096
  disk_size: 25
  local_disk: True
  hourly_billing: True
  domain: MEL.SL.9SPOKES.COM
  location: mel01
  # Optional
  max_net_speed: 1000
  public_vlan: 947935
  private_vlan: 712122
  private_ssh: True
```

And the log of the `salt-cloud` run.

```
root@MGT01:~# salt-cloud -p web_proxy_mel_sl test1

Warning: Permanently added '10.118.7.249' (ECDSA) to the list of known hosts.
Password:
Sun Sep 13 19:14:51 CDT 2015
Warning: Permanently added '10.118.7.249' (ECDSA) to the list of known hosts.
Password: Connection to 10.118.7.249 closed.

Warning: Permanently added '10.118.7.249' (ECDSA) to the list of known hosts.
Password: Connection to 10.118.7.249 closed.

Warning: Permanently added '10.118.7.249' (ECDSA) to the list of known hosts.
Password: Connected to 10.118.7.249.

sftp> put  /tmp/tmphBrdXf /tmp/.saltcloud-3ea33a42-e333-4343-8774-18413cb350b7/minion.pem
Uploading /tmp/tmphBrdXf to /tmp/.saltcloud-3ea33a42-e333-4343-8774-18413cb350b7/minion.pem
/tmp/tmphBrdXf                                                                                                                                   100% 1674     1.6KB/s   00:00
Warning: Permanently added '10.118.7.249' (ECDSA) to the list of known hosts.
Password: Connection to 10.118.7.249 closed.

Warning: Permanently added '10.118.7.249' (ECDSA) to the list of known hosts.
Password: Connected to 10.118.7.249.

sftp> put  /tmp/tmpXwWwBT /tmp/.saltcloud-3ea33a42-e333-4343-8774-18413cb350b7/minion.pub
Uploading /tmp/tmpXwWwBT to /tmp/.saltcloud-3ea33a42-e333-4343-8774-18413cb350b7/minion.pub
/tmp/tmpXwWwBT                                                                                                                                   100%  450     0.4KB/s   00:00
Warning: Permanently added '10.118.7.249' (ECDSA) to the list of known hosts.
Password: Connected to 10.118.7.249.

sftp> put  /tmp/tmpeeJZ90 /tmp/.saltcloud-3ea33a42-e333-4343-8774-18413cb350b7/minion
Uploading /tmp/tmpeeJZ90 to /tmp/.saltcloud-3ea33a42-e333-4343-8774-18413cb350b7/minion
/tmp/tmpeeJZ90                                                                                                                                   100% 4516     4.4KB/s   00:00
Warning: Permanently added '10.118.7.249' (ECDSA) to the list of known hosts.
Password: Connection to 10.118.7.249 closed.

Warning: Permanently added '10.118.7.249' (ECDSA) to the list of known hosts.
Password: Connection to 10.118.7.249 closed.

Warning: Permanently added '10.118.7.249' (ECDSA) to the list of known hosts.
Password: Connection to 10.118.7.249 closed.

Warning: Permanently added '10.118.7.249' (ECDSA) to the list of known hosts.
Password: Connection to 10.118.7.249 closed.

[INFO    ] Salt installed on test1
test1:
    ----------
    accountId:
        314306
    createDate:
        2015-09-14T12:09:06+12:00
    deployed:
        True
    domain:
        MEL.SL.9SPOKES.COM
    fullyQualifiedDomainName:
        test1.MEL.SL.9SPOKES.COM
    globalIdentifier:
        443aa07a-d0f3-40b4-863e-d959a133a7bd
    hostname:
        test1
    id:
        12329659
    lastPowerStateId:
    lastVerifiedDate:
    maxCpu:
        2
    maxCpuUnits:
        CORE
    maxMemory:
        4096
    metricPollDate:
    modifyDate:
    password:
        N99wYdC7
    provisionDate:
    public_ip:
        10.118.7.249
    startCpus:
        2
    statusId:
        1001
    username:
        root
    uuid:
        20cff431-9251-413a-a0df-68c5353076c6
```

**Comment by rallytime** [comment]
@geraint-jones-nz Sorry I should have been more clear. Can you specify what you're doing to see that salt isn't installed? If you log into the minion, there isn't a salt package? Or the minion just isn't running? Do you have the minion pointed to the master in your cloud provider file correctly (I don't see it in the profile you posted)?

I can't reproduce this issue, so the more details you can give, the better. Thanks!

**Comment by geraint-jones-nz** [comment]
That is correct, when I ssh into the host there is no minion installed. And no /etc/salt directory.

Geraint Jones - Systems and Network Consultant
P: 022 123 4626 | T: @nexus_monkey_nz | W: geraintjones.co.nz

**Comment by rallytime** [comment]
Ok. Can you please post your **sanitized** cloud configurations files? Both the provider and profile would be great, just be sure to remove the sensitive bits.

**Comment by rallytime** [comment]
Actually, I just realized you did post your profile. Though the provider file is still useful so we can see what other configs you have set. Is there anything in the debug logs that might be useful?

I am wondering if something wrong might be going on with the bootstrap script. Can you try running this again, but with `-l debug` on the single minion so we can see if any stacktraces or anything like that is being triggered?

I also see that you're deploying an Ubuntu image. Can you also try adding `script_args: -Z -P` to your profile? (That will allow pip-based installations as well as an updated version of ZMQ. These arguments are required on newer versions of the bootstrap script.)

**Comment by geraint-jones-nz** [comment]
I have added `script_args: -Z -P` and nothing changed.

Here is the `cloud.providers`

```
9s-softlayer:
    minion:
      master: <private ip of master>
    user: geraint_jones
    apikey: '<snip>'
    provider: softlayer
    ssh_connect_timeout: 1800
```


Here is the pertinent bits from the logs produced with `-l debug` (let me know if you want the whole transcript with SL API calls)

```
[DEBUG   ] MasterEvent PUB socket URI: ipc:///var/run/salt/master/master_event_pub.ipc
[DEBUG   ] MasterEvent PULL socket URI: ipc:///var/run/salt/master/master_event_pull.ipc
[DEBUG   ] Sending event - data = {'_stamp': '2015-09-15T01:29:53.422242', 'event': 'executing deploy script', 'kwargs': {'tty': True, 'sudo': False, 'start_action': None, 'make_minion': True, 'port': '22', 'deploy_command': '/tmp/.saltcloud/deploy.sh', 'script': '', 'conf_file': '/etc/salt/master', 'maxtries': 15, 'username': 'root', 'key_filename': None, 'tmp_dir': '/tmp/.saltcloud', 'keep_tmp': False, 'minion_conf': {'ioflo_realtime': True, 'master_alive_interval': 0, 'recon_default': 1000, 'whitelist_modules': [], 'tcp_keepalive_intvl': -1, 'utils_dirs': [], 'state_top_saltenv': None, 'states_dirs': [], 'state_output_diff': False, 'fileserver_backend': ['roots'], 'outputter_dirs': [], 'raet_port': 4510, 'module_dirs': [], 'git_pillar_ssl_verify': False, 'gitfs_insecure_auth': False, 'state_auto_order': True, 'acceptance_wait_time': 10, 'disable_modules': [], 'backup_mode': '', 'cluster_masters': [], 'recon_randomize': True, 'file_ignore_glob': None, 'auto_accept': True, 'open_mode': False, 'winrepo_ssl_verify': False, 'git_pillar_pubkey': '', 'state_verbose': True, 'verify_master_pubkey_sign': False, 'password': None, 'cluster_mode': False, 'startup_states': '', 'grains': {}, 'state_output': 'full', 'tcp_pull_port': 4511, 'gitfs_pubkey': '', 'fileserver_ignoresymlinks': False, 'retry_dns': 30, 'pillar_roots': {'base': ['/srv/pillar', '/srv/spm/pillar']}, 'master_shuffle': False, 'multiprocessing': True, 'file_roots': {'base': ['/srv/salt', '/srv/spm/salt']}, 'top_file_merging_strategy': 'merge', 'root_dir': '/', 'log_granular_levels': {}, 'git_pillar_insecure_auth': False, 'returner_dirs': [], 'gitfs_privkey': '', 'tcp_keepalive': True, 'grains_dirs': [], 'state_top': 'top.sls', 'transport': 'zeromq', 'gitfs_base': 'master', 'render_dirs': [], 'gitfs_user': '', 'fileserver_limit_traversal': False, 'ioflo_console_logdir': '', 'conf_file': '/etc/salt/minion', 'top_file': '', 'zmq_monitor': False, 'file_recv_max_size': 100, 'pidfile': '/var/run/salt-minion.pid', 'range_server': 'range:80', 'raet_mutable': False, 'log_datefmt_logfile': '%Y-%m-%d %H:%M:%S', 'file_ignore_regex': None, 'winrepo_cachefile': 'winrepo.p', 'winrepo_insecure_auth': False, 'raet_main': False, 'cachedir': '/var/cache/salt/minion', 'winrepo_passphrase': '', 'interface': '0.0.0.0', 'update_restart_services': [], 'recon_max': 10000, 'master_uri_format': 'default', 'default_include': 'minion.d/*.conf', 'environment': None, 'http_request_timeout': 3600.0, 'rejected_retry': False, 'state_events': False, 'gitfs_ssl_verify': False, 'http_max_body': 107374182400, 'restart_on_error': False, 'winrepo_dir': '/srv/salt/win/repo', 'raet_road_bufcnt': 2, 'winrepo_pubkey': '', 'ipc_mode': 'ipc', 'keysize': 2048, 'master_sign_key_name': 'master_sign', 'cython_enable': False, 'sls_list': [], 'git_pillar_env': '', 'ext_job_cache': '', 'verify_env': True, 'hash_type': 'md5', 'winrepo_source_dir': 'salt://win/repo-ng/', 'always_verify_signature': False, 'modules_max_memory': -1, 'renderer': 'yaml_jinja', 'random_reauth_delay': 10, 'gitfs_env_whitelist': [], 'git_pillar_base': 'master', 'gitfs_passphrase': '', 'extension_modules': '', 'mine_interval': 60, 'git_pillar_privkey': '', 'winrepo_dir_ng': '/srv/salt/win/repo-ng', 'minion_id_caching': True, 'winrepo_user': '', 'grains_cache': False, 'file_recv': False, 'log_level_logfile': 'info', 'ipv6': False, 'master': '10.118.7.200', 'sudo_user': '', 'username': None, 'master_finger': '', 'failhard': False, 'tcp_keepalive_idle': 300, 'winrepo_privkey': '', 'log_fmt_logfile': '%(asctime)s,%(msecs)03.0f [%(name)-17s][%(levelname)-8s][%(process)d] %(message)s', 'master_port': '4506', 'enable_zip_modules': False, 'ioflo_period': 0.1, 'use_master_when_local': False, 'ping_interval': 0, 'auth_timeout': 60, 'tcp_keepalive_cnt': -1, 'raet_lane_bufcnt': 100, 'git_pillar_root': '', 'winrepo_remotes_ng': ['https://github.com/saltstack/salt-winrepo-ng.git'], 'raet_alt_port': 4511, 'env_order': [], 'state_aggregate': False, 'syndic_log_file': '/var/log/salt/syndic', 'update_url': False, 'grains_refresh_every': 0, 'winrepo_password': '', 'git_pillar_branch': 'master', 'default_top': 'base', 'winrepo_branch': 'master', 'providers': {}, 'winrepo_remotes': ['https://github.com/saltstack/salt-winrepo.git'], 'autoload_dynamic_modules': True, 'file_buffer_size': 262144, 'log_fmt_console': '[%(levelname)-8s] %(message)s', 'random_master': False, 'log_datefmt': '%H:%M:%S', 'grains_cache_expiration': 300, 'minion_floscript': '/usr/lib/python2.7/dist-packages/salt/config/daemons/flo/minion.flo', 'id': 'test5', 'caller_floscript': '/usr/lib/python2.7/dist-packages/salt/config/daemons/flo/caller.flo', 'loop_interval': 1, 'log_level': 'info', 'git_pillar_password': '', 'git_pillar_user': '', 'auth_safemode': False, 'clean_dynamic_modules': True, 'disable_returners': [], 'cache_sreqs': True, 'autosign_timeout': 120, 'gitfs_root': '', 'test': False, 'gitfs_password': '', 'syndic_pidfile': '/var/run/salt-syndic.pid', 'syndic_finger': '', 'raet_clear_remotes': True, 'raet_clear_remote_masters': True, 'file_client': 'remote', 'user': 'root', 'git_pillar_passphrase': '', 'pillarenv': None, 'auth_tries': 7, 'cache_jobs': False, 'permissive_pki_access': False, 'cmd_safe': True, 'gitfs_env_blacklist': [], 'zmq_filtering': False, 'master_type': 'str', 'pki_dir': '/etc/salt/pki/minion', 'max_event_size': 1048576, 'ioflo_verbose': 0, 'sock_dir': '/var/run/salt/minion', 'tcp_pub_port': 4510, 'acceptance_wait_time_max': 0, 'fileserver_followsymlinks': True, 'log_file': '/var/log/salt/minion', 'gitfs_remotes': [], 'gitfs_mountpoint': ''}, 'file_map': None, 'host': '10.118.7.213', 'preseed_minion_keys': None, 'inline_script': None, 'parallel': False, 'display_ssh_output': True, 'name': 'test5', 'script_env': None, 'has_ssh_agent': False, 'known_hosts_file': '/dev/null', 'sock_dir': '/var/run/salt/master', 'script_args': '-Z -P', 'salt_host': '10.118.7.213', 'master_sign_pub_file': None}}
[DEBUG   ] Deploying 10.118.7.213 at 1442280593.0
[DEBUG   ] Attempting connection to host 10.118.7.213 on port 22
[DEBUG   ] SSH port 22 on 10.118.7.213 is available
[DEBUG   ] Using password authentication
[DEBUG   ] Attempting to authenticate as root (try 1 of 15)
[DEBUG   ] SSH command: 'ssh -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -oConnectTimeout=15 -p 22 root@10.118.7.213 date'
[DEBUG   ] Child Forked! PID: 27163  STDOUT_FD: 6  STDERR_FD: 8
[DEBUG   ] Terminal Command: /bin/sh -c ssh -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -oConnectTimeout=15 -p 22 root@10.118.7.213 date
Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
[DEBUG   ] Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
Password: [DEBUG   ] Password:

Mon Sep 14 20:29:34 CDT 2015
[DEBUG   ] Mon Sep 14 20:29:34 CDT 2015
[DEBUG   ] Logging into 10.118.7.213:22 as root
[DEBUG   ] Using Qx9m7QNd as the password
[DEBUG   ] SSH command: 'ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -p 22 root@10.118.7.213 'test -e '"'"'/tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a'"'"'''
[DEBUG   ] Child Forked! PID: 27165  STDOUT_FD: 6  STDERR_FD: 8
[DEBUG   ] Terminal Command: /bin/sh -c ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -p 22 root@10.118.7.213 'test -e '"'"'/tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a'"'"''
Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
[DEBUG   ] Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
Password: [DEBUG   ] Password:
Connection to 10.118.7.213 closed.
[DEBUG   ] Connection to 10.118.7.213 closed.

[DEBUG   ] SSH command: 'ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -p 22 root@10.118.7.213 'sh -c "( mkdir -p '"'"'/tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a'"'"' && chmod 700 '"'"'/tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a'"'"' )"''
[DEBUG   ] Child Forked! PID: 27167  STDOUT_FD: 6  STDERR_FD: 8
[DEBUG   ] Terminal Command: /bin/sh -c ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -p 22 root@10.118.7.213 'sh -c "( mkdir -p '"'"'/tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a'"'"' && chmod 700 '"'"'/tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a'"'"' )"'
Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
[DEBUG   ] Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
Password: [DEBUG   ] Password:
Connection to 10.118.7.213 closed.
[DEBUG   ] Connection to 10.118.7.213 closed.

[DEBUG   ] Uploading /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pem to 10.118.7.213 (sfcp)
[DEBUG   ] SFTP command: 'echo "put  /tmp/tmpQW4cmV /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pem" | sftp -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -oPort=22 root@10.118.7.213'
[DEBUG   ] Child Forked! PID: 27169  STDOUT_FD: 7  STDERR_FD: 9
[DEBUG   ] Terminal Command: /bin/sh -c echo "put  /tmp/tmpQW4cmV /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pem" | sftp -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -oPort=22 root@10.118.7.213
Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
[DEBUG   ] Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
Password: [DEBUG   ] Password:
Connected to 10.118.7.213.
[DEBUG   ] Connected to 10.118.7.213.

sftp> put  /tmp/tmpQW4cmV /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pem
Uploading /tmp/tmpQW4cmV to /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pem
/tmp/tmpQW4cmV                                                                                                                                   100% 1678     1.6KB/s   00:00
[DEBUG   ] sftp> put  /tmp/tmpQW4cmV /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pem
Uploading /tmp/tmpQW4cmV to /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pem
/tmp/tmpQW4cmV                                                                                                                                   100% 1678     1.6KB/s   00:00 ETA
[DEBUG   ] SSH command: 'ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -p 22 root@10.118.7.213 'chmod 600 '"'"'/tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pem'"'"'''
[DEBUG   ] Child Forked! PID: 27173  STDOUT_FD: 7  STDERR_FD: 9
[DEBUG   ] Terminal Command: /bin/sh -c ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -p 22 root@10.118.7.213 'chmod 600 '"'"'/tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pem'"'"''
Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
[DEBUG   ] Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
Password: [DEBUG   ] Password:
Connection to 10.118.7.213 closed.
[DEBUG   ] Connection to 10.118.7.213 closed.

[DEBUG   ] Uploading /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pub to 10.118.7.213 (sfcp)
[DEBUG   ] SFTP command: 'echo "put  /tmp/tmpO4U0SL /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pub" | sftp -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -oPort=22 root@10.118.7.213'
[DEBUG   ] Child Forked! PID: 27175  STDOUT_FD: 8  STDERR_FD: 10
[DEBUG   ] Terminal Command: /bin/sh -c echo "put  /tmp/tmpO4U0SL /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pub" | sftp -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -oPort=22 root@10.118.7.213
Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
[DEBUG   ] Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
Password: [DEBUG   ] Password:
Connected to 10.118.7.213.
[DEBUG   ] Connected to 10.118.7.213.

sftp> put  /tmp/tmpO4U0SL /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pub
Uploading /tmp/tmpO4U0SL to /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pub
/tmp/tmpO4U0SL                                                                                                                                   100%  450     0.4KB/s   00:00
[DEBUG   ] sftp> put  /tmp/tmpO4U0SL /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pub
Uploading /tmp/tmpO4U0SL to /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pub
/tmp/tmpO4U0SL                                                                                                                                   100%  450     0.4KB/s   00:00 ETA
[DEBUG   ] Uploading /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion to 10.118.7.213 (sfcp)
[DEBUG   ] SFTP command: 'echo "put  /tmp/tmpGikCOR /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion" | sftp -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -oPort=22 root@10.118.7.213'
[DEBUG   ] Child Forked! PID: 27179  STDOUT_FD: 9  STDERR_FD: 11
[DEBUG   ] Terminal Command: /bin/sh -c echo "put  /tmp/tmpGikCOR /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion" | sftp -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -oPort=22 root@10.118.7.213
Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
[DEBUG   ] Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
Password: [DEBUG   ] Password:
Connected to 10.118.7.213.
[DEBUG   ] Connected to 10.118.7.213.

sftp> put  /tmp/tmpGikCOR /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion
Uploading /tmp/tmpGikCOR to /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion
/tmp/tmpGikCOR                                                                                                                                   100% 4517     4.4KB/s   00:00
[DEBUG   ] sftp> put  /tmp/tmpGikCOR /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion
Uploading /tmp/tmpGikCOR to /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion
/tmp/tmpGikCOR                                                                                                                                   100% 4517     4.4KB/s   00:00 ETA
[DEBUG   ] SSH command: 'ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -p 22 root@10.118.7.213 'rm -f '"'"'/tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pub'"'"'''
[DEBUG   ] Child Forked! PID: 27183  STDOUT_FD: 9  STDERR_FD: 11
[DEBUG   ] Terminal Command: /bin/sh -c ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -p 22 root@10.118.7.213 'rm -f '"'"'/tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pub'"'"''
Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
[DEBUG   ] Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
Password: [DEBUG   ] Password:
Connection to 10.118.7.213 closed.
[DEBUG   ] Connection to 10.118.7.213 closed.

[DEBUG   ] Removed /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pub
[DEBUG   ] SSH command: 'ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -p 22 root@10.118.7.213 'rm -f '"'"'/tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pem'"'"'''
[DEBUG   ] Child Forked! PID: 27185  STDOUT_FD: 9  STDERR_FD: 11
[DEBUG   ] Terminal Command: /bin/sh -c ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -p 22 root@10.118.7.213 'rm -f '"'"'/tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pem'"'"''
Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
[DEBUG   ] Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
Password: [DEBUG   ] Password:
Connection to 10.118.7.213 closed.
[DEBUG   ] Connection to 10.118.7.213 closed.

[DEBUG   ] Removed /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion.pem
[DEBUG   ] SSH command: 'ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -p 22 root@10.118.7.213 'rm -f '"'"'/tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/grains'"'"'''
[DEBUG   ] Child Forked! PID: 27187  STDOUT_FD: 9  STDERR_FD: 11
[DEBUG   ] Terminal Command: /bin/sh -c ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -p 22 root@10.118.7.213 'rm -f '"'"'/tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/grains'"'"''
Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
[DEBUG   ] Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
Password: [DEBUG   ] Password:
Connection to 10.118.7.213 closed.
[DEBUG   ] Connection to 10.118.7.213 closed.

[DEBUG   ] Removed /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/grains
[DEBUG   ] SSH command: 'ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -p 22 root@10.118.7.213 'rm -f '"'"'/tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion'"'"'''
[DEBUG   ] Child Forked! PID: 27189  STDOUT_FD: 9  STDERR_FD: 11
[DEBUG   ] Terminal Command: /bin/sh -c ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -p 22 root@10.118.7.213 'rm -f '"'"'/tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion'"'"''
Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
[DEBUG   ] Warning: Permanently added '10.118.7.213' (ECDSA) to the list of known hosts.
Password: [DEBUG   ] Password:
Connection to 10.118.7.213 closed.
[DEBUG   ] Connection to 10.118.7.213 closed.

[DEBUG   ] Removed /tmp/.saltcloud-f619439d-fd5b-4899-918b-90199602374a/minion
[DEBUG   ] MasterEvent PUB socket URI: ipc:///var/run/salt/master/master_event_pub.ipc
[DEBUG   ] MasterEvent PULL socket URI: ipc:///var/run/salt/master/master_event_pull.ipc
[DEBUG   ] Sending event - data = {'host': '10.118.7.213', 'name': 'test5', '_stamp': '2015-09-15T01:30:17.891802', 'event': 'test5 has been deployed at 10.118.7.213'}
[INFO    ] Salt installed on test5
[DEBUG   ] MasterEvent PUB socket URI: ipc:///var/run/salt/master/master_event_pub.ipc
[DEBUG   ] MasterEvent PULL socket URI: ipc:///var/run/salt/master/master_event_pull.ipc
[DEBUG   ] Sending event - data = {'profile': 'web_proxy_mel_sl', 'event': 'created instance', '_stamp': '2015-09-15T01:30:17.943797', 'name': 'test5', 'provider': '9s-softlayer:softlayer'}
[DEBUG   ] LazyLoaded nested.output
```

**Comment by rallytime** [comment]
Unfortunately, I still cannot reproduce this error and I've tried many different combinations. I am starting to wonder if it may have something to do with the version of salt you're using. I can't seem to locate the sha from the output of your versions report in salt, so I can't test that. I did, however, replicate your config files, replacing your key/user/etc values with my own, and everything worked very well.

Previous to this comment, my master was running outside of softlayer. I made a new master inside softlayer in order to make sure to use the configs you were using, especially `private_ssh: True`. I installed the master using the bootstrap script so I could install from the 2015.8.0. I'll post all of my reproduction steps and configs here so you can maybe try them. Otherwise I am not sure what else to try.

Install salt-master via bootstrap, without a minion at version 2015.8.0:
```
curl -L https://bootstrap.saltstack.com -o install_salt.sh
sh install_salt -Z -P -M -N git v2015.8.0
```
Install softlayer dependency:
```
pip install softlayer
```
Make cloud provider and profile configs (note the script_args in the profile):
```
# /etc/salt/cloud.providers
softlayer-config:
  minion:
    master: <my-master's-private-ip>
  user: <redacted>
  apikey: <redacted>
  provider: softlayer
  ssh_connect_timeout: 1800

# /etc/salt/cloud.profiles
sl-ubuntu-14:
  provider: softlayer-config
  image: UBUNTU_14_64
  cpu_number: 1
  ram: 1024
  disk_size: 100
  local_disk: True
  hourly_billing: True
  domain: example.com
  location: sjc01
  script_args: '-P -Z git v2015.8.0'
  max_net_speed: 1000
  public_vlan: 988609
  private_vlan: 988611
  private_ssh: True
```
Then run the cloud command:
```
# salt-cloud -p sl-ubuntu-14 nt-1
```
Which returns successfully and I can ping the minion:
```
# salt nt-1 test.ping
nt-1:
    True
```

For completeness, here are the version reports for each:

Master:
```
# salt-cloud --versions
Salt Version:
            Salt: 2015.8.0

Dependency Versions:
 Apache Libcloud: Not Installed
          Jinja2: 2.7.2
        M2Crypto: 0.21.1
            Mako: Not Installed
          PyYAML: 3.10
           PyZMQ: 14.5.0
          Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
            RAET: Not Installed
         Tornado: 4.2.1
             ZMQ: 4.0.5
            cffi: 0.8.6
        cherrypy: Not Installed
        dateutil: 1.5
           gitdb: Not Installed
       gitpython: Not Installed
           ioflo: Not Installed
         libnacl: Not Installed
    msgpack-pure: Not Installed
  msgpack-python: 0.3.0
    mysql-python: Not Installed
       pycparser: 2.10
        pycrypto: 2.6.1
          pygit2: Not Installed
    python-gnupg: Not Installed
           smmap: Not Installed
         timelib: Not Installed

System Versions:
            dist: Ubuntu 14.04 trusty
         machine: x86_64
         release: 3.13.0-55-generic
          system: Ubuntu 14.04 trusty
```
Minion:
```
# salt nt-1 test.versions_report
nt-1:
    Salt Version:
               Salt: 2015.8.0

    Dependency Versions:
             Jinja2: 2.7.2
           M2Crypto: 0.21.1
               Mako: Not Installed
             PyYAML: 3.10
              PyZMQ: 14.5.0
             Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
               RAET: Not Installed
            Tornado: 4.2.1
                ZMQ: 4.0.5
               cffi: 0.8.6
           cherrypy: Not Installed
           dateutil: Not Installed
              gitdb: Not Installed
          gitpython: Not Installed
              ioflo: Not Installed
            libnacl: Not Installed
       msgpack-pure: Not Installed
     msgpack-python: 0.3.0
       mysql-python: Not Installed
          pycparser: 2.10
           pycrypto: 2.6.1
             pygit2: Not Installed
       python-gnupg: Not Installed
              smmap: Not Installed
            timelib: Not Installed

    System Versions:
               dist: Ubuntu 14.04 trusty
            machine: x86_64
            release: 3.13.0-55-generic
             system: Ubuntu 14.04 trusty
```

So, that all being said, @geraint-jones-nz Is there a setting that I am missing? If not, I'd try to see if getting a newer version of the salt-master installed and then making sure the minion has those script args to match, based on the steps I've outlined above, and see if that will work for you.

**Comment by hexedpackets** [comment]
Is this a new master or an in-place upgrade? I recently saw the same issue from a new salt master. Turns out I needed to run `salt-cloud -u` to pull down the bootstrap script. By default salt-cloud looks in /etc/salt/cloud.deploy.d/ for bootstrap scripts, with the default name of 'bootstrap-salt', and if doesn't find it it continues without running anything on the minion.

**Comment by rallytime** [comment]
Oh interesting...that makes more sense. How were you installing the salt-master? And which OS does your master reside on?

The latest (internal to softlayer) master I used the commands posted above to install the master via salt bootstrap.

**Comment by rallytime** [comment]
@hexedpackets Did you see this when using softlayer or a different driver?

**Comment by geraint-jones-nz** [comment]
Just testing. Will let yall know asap.

**Comment by geraint-jones-nz** [comment]
@hexedpackets where do I send the beer ?

So it is now working, and as you surmised was an inplace upgrade.

**Comment by rallytime** [comment]
@hexedpackets FTW! 

@geraint-jones-nz Yay! Glad it is working for you now. Which version were you upgrading from? And which OS is your master on?

**Comment by geraint-jones-nz** [comment]
v2015.5 to v2015.8 using `bootsrap.sh`

**Comment by hexedpackets** [comment]
@geraint-jones-nz Awesome, glad that worked for you too!

@rallytime I saw this happening with both the ec2 and saltify drivers. I installed with pip so obviously the /etc/salt files aren't automatically setup. I think that's ok but It would be nice to have this behaviour documented somewhere. I had to dig into the salt.utils.cloud code to figure out where it was looking since the cloud config param `deploy_scripts_search_path` isn't mentioned anywhere.

**Comment by rallytime** [comment]
@hexedpackets - Yes good call. We should definitely get that documented.

I am curious though, how @geraint-jones-nz ended up in this state if he _wasn't_ using pip to install. I'd like to track this down and figure out how that happened in case there is a deeper problem here.

@geraint-jones-nz Can you tell me what the output of `ls /usr/lib/python2.7/dist-packages/salt/cloud/deploy/` is? Do you see a `bootstrap-salt.sh` file in that directory?

**Comment by rallytime** [comment]
I think there is something deeper going on here, but I can't quite figure out what it is, because I can't reproduce it. You shouldn't **have** to have the `/etc/salt/cloud.deploy.d/` directory for salt-cloud to work. It should be able to pull it from `/path/to/installed/salt/cloud/deploy/bootstrap-salt.sh` even if you installed via pip or the bootstrap script. I've tried doing just that on a fresh VM and it is working just fine. 

For example, I installed with pip and the bootstrap file is correctly in the directory:
```
root@nt1:~# pip show salt
---
Name: salt
Version: 2015.8.0
Location: /usr/local/lib/python2.7/dist-packages
Requires: Jinja2, msgpack-python, PyYAML, MarkupSafe, requests, tornado, futures, pycrypto, pyzmq
root@nt1:~# ls /usr/local/lib/python2.7/dist-packages/salt/cloud/deploy/
Arch-git.sh            curl-bootstrap.sh  Fedora.sh       python-bootstrap.sh  RHEL6.sh       wget-bootstrap-nocert.sh
Arch.sh                Debian-git.sh      FreeBSD-git.sh  RHEL5-git.sh         SmartOS.sh     wget-bootstrap.sh
bootstrap-salt.sh      Debian.sh          FreeBSD.sh      RHEL5.sh             Ubuntu-git.sh
curl-bootstrap-git.sh  Fedora-git.sh      None.sh         RHEL6-git.sh         Ubuntu.sh
```
And the `/etc/salt/` directory does **not** have the `/etc/salt/cloud.deploy.d/` directory:
```
root@nt1:~# ls /etc/salt/
cloud.profiles.d  cloud.providers.d  minion  minion.d  minion_id  pki
```
But I can successfully create a minion with salt-cloud on SoftLayer and EC2. I will have to keep thinking about this, as I can't seem to figure out where this is falling down for you guys. (Though I am glad the `salt-cloud -u` has solved the problem for now.)

**Comment by centromere** [comment]
With a brand new master setup (including the ec2.py and `salt-cloud -u` fixes mentioned above), I get the following error:

    [INFO    ] Rendering deploy script: /etc/salt/cloud.deploy.d/bootstrap-salt.sh
    [ERROR   ] Failed to deploy 'db02'. Error: Command 'ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -oPasswordAuthentication=no -oChallengeResponseAuthentication=no -oPubkeyAuthentication=yes -oKbdInteractiveAuthentication=no -i /etc/salt/salt-ssh.key -p 22 ubuntu@10.0.4.88 \'sudo /tmp/.saltcloud-00363c15-e8a8-46d0-a34b-2fb33f6dc9e9/deploy.sh -c \'"\'"\'/tmp/.saltcloud-00363c15-e8a8-46d0-a34b-2fb33f6dc9e9\'"\'"\'\'' failed. Exit code: 1

The command executed was: `salt-cloud -m /srv/config/cloud.maps.d/all.map -P`

    Salt Version:
               Salt: 2015.8.0
 
    Dependency Versions:
             Jinja2: 2.7.2
           M2Crypto: Not Installed
               Mako: 0.9.1
             PyYAML: 3.10
              PyZMQ: 14.4.0
             Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
               RAET: Not Installed
            Tornado: 4.2.1
                ZMQ: 4.0.4
               cffi: Not Installed
           cherrypy: Not Installed
           dateutil: 1.5
              gitdb: 0.5.4
          gitpython: Not Installed
              ioflo: Not Installed
            libnacl: Not Installed
       msgpack-pure: Not Installed
     msgpack-python: 0.3.0
       mysql-python: Not Installed
          pycparser: Not Installed
           pycrypto: 2.6.1
             pygit2: Not Installed
       python-gnupg: Not Installed
              smmap: 0.8.2
            timelib: Not Installed
 
    System Versions:
               dist: Ubuntu 14.04 trusty
            machine: x86_64
            release: 3.13.0-65-generic
             system: Ubuntu 14.04 trusty

**Comment by rallytime** [comment]
@centromere - This looks like a different error/failure to me. Would you mind creating a new issue with any more information you can provide about this instance that is failing?

If you could post your relevant cloud provider, profile, and map files (be sure remove any sensitive information) as well if you're seeing problems just with one particular instance or the entire map file failing, etc etc, that would be very helpful.

**Comment by centromere** [comment]
@rallytime I deleted my comment above because I was able to fix the issue by passing "-P" to the bootstrap script via script_args.

**Comment by rallytime** [comment]
:+1:

**Comment by rallytime** [comment]
@hexedpackets When you were tracking this down, was your `deploy_scripts_search_path` value empty in `opts`??

I've traced this back to the `salt.utils.cloud.os_script` function [returning an empty string](https://github.com/saltstack/salt/blob/9a0171089d18b5f678810ba59478de5361bfcb2d/salt/utils/cloud.py#L147). The only way that would happen is if that opts value doesn't have a path in it...which is...weird.

I still haven't been able to reproduce this issue unfortunately, but from the awesome logs and conversations posted here, that's definitely where the `script` arg isn't getting passed through to the `salt.utils.cloud.bootstrap` function correctly in the [deploy_kwargs](https://github.com/saltstack/salt/blob/9a0171089d18b5f678810ba59478de5361bfcb2d/salt/utils/cloud.py#L366).

**Comment by hexedpackets** [comment]
@rallytime Interesting. I see a different path loaded:

```python
In [1]: import salt.config
In [2]: opts = salt.config.cloud_config('/etc/salt/cloud')
In [3]: opts['deploy_scripts_search_path']
Out[3]:
('/etc/salt/cloud.deploy.d',
 '/usr/lib/python2.7/site-packages/salt/config/cloud/deploy')
```

That path (/usr/lib/python2.7/site-packages/salt/config/cloud/deploy) does not exist, but /usr/lib/python2.7/site-packages/salt/cloud/deploy does.

**Comment by bogdanr** [comment]
Also, when you install salt with pip from git (`pip install -e ~/salt`) that path should be defined in /etc/salt/cloud with the `deploy_scripts_search_path: /root/salt/salt/cloud/deploy` option

**Comment by rallytime** [comment]
The next thing I was going to look at was to see if something is getting crunched in `setup.py`, but I didn't make it back around to this one yet.

**Comment by rallytime** [comment]
We were finally able to reproduce this one, and the fix is in #31172. :D

**Comment by rallytime** [comment]
Since we were able to reproduce this and confirm the fix, I'm going to close this. Thanks everyone!

**Comment by cconrad** [comment]
Hi, not used to reporting bugs on GitHub, my apologies if this is the wrong place to comment.

As far as I understand, the issue here was 1) it should not be necessary to run "salt-cloud -u" after installing and before provisioning VMs and 2) should have been fixed in v2015.8.8. If that is not correct, please correct me and disregard the remainder :)

I installed salt v2015.8.8 on FreeBSD 10.1 (not an upgrade) as follows:

```
sudo pkg update && pkg upgrade –y 
fetch -o bootstrap_salt.sh https://bootstrap.saltstack.com 
sudo sh bootstrap_salt.sh -M -L stable 
```

Output from `pkg info | grep salt`:
```
py27-salt-2015.8.8_1           Distributed remote execution and configuration management system
```

Output from `salt --versions`:
```
Salt Version:
           Salt: 2015.8.8
 
Dependency Versions:
         Jinja2: 2.8
       M2Crypto: Not Installed
           Mako: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.2.0
         Python: 2.7.11 (default, Mar 19 2016, 01:13:51)
           RAET: Not Installed
        Tornado: 4.3
            ZMQ: 4.1.4
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.5.0
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
        libgit2: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.7
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: Not Installed
        timelib: Not Installed
 
System Versions:
           dist:   
        machine: amd64
        release: 10.1-RELEASE-p31
```

Output from `cat /usr/local/etc/salt/cloud`:
```
delete_sshkeys: True
minion:
  master: <redacted>
```

Output from `cat /usr/local/etc/salt/cloud.providers.d/gce.conf`:
```
gce:
  project: <redacted>
  service_account_email_address: <redacted>
  service_account_private_key: <redacted>
  driver: gce
```

Output from `cat /usr/local/etc/salt/cloud.profiles.d/gce-freebsd.conf`:
```
gce-freebsd:
  image: "https://www.googleapis.com/compute/v1/projects/freebsd-org-cloud-dev/global/images/freebsd-10-3-release-amd64"
  size: n1-standard-2
  location: europe-west1-b
  network: default
  use_persistent_disk: True
  delete_boot_pd: True
  ssh_interface: public_ips
  provider: gce
  ssh_username: freebsd
  ssh_keyfile: <redacted>
```

I then ran `sudo salt-cloud -l debug -p gce-freebsd test-salt-iocage --keep-tmp` and experienced the issue from this thread:
1. The instance is created successfully.
2. The SSH connection works fine.
3. The files "minion", "minion.pem" and "minion.pub" are copied to the instance.
4. The next line in the output is `[INFO    ] Salt installed on test-salt-iocage`. No attempt is made to copy or run the bootstrap script.
5. When connecting to the new instance, no salt-related packages are installed on it.

The solution provided by @hexedpackets, to run `sudo salt-cloud -u`, created the directory `/usr/local/etc/salt/cloud.deploy.d` with the file `bootstrap-salt.sh` inside and solved the issue for me, but shouldn't it have been unnecessary on this version?

**Comment by rallytime** [comment]
@cconrad You have explained the issue correctly, and you're indeed running into this bug. The fix I mentioned above fixed this bug on the 2016.3 branch, which I backported to the 2015.8 branch. Unfortunately, I caused a regression listed in #32183 (and now you're running into here). The fix for that regression, #32105, was submitted to the 2015.8 branch (and is confirmed tested and fixed), but unfortunately was merged after we cut 2015.8.8. I wrote a test to catch this regression in the future with #32295.

So, in summary, this is fixed at the HEAD of the 2015.8 branch, will be available in the 2015.8.9 release. For now, the `salt-cloud -u` is still needed. Apologies for the confusion.
